### ISSUE TITLE
on('notification') not firing when tapping on notification while app is in background

### ISSUE THREAD

**Issue by mdbw3** [phonegap/phonegap-plugin-push#1033]
### Expected Behaviour

on('notification') is fired when a notification is tapped while app is in background
### Actual Behaviour

on('notification') isn't fired
#### Steps to Reproduce

Open app, push on windows button (app goes to background), send notification, tap on notification -> app opens, but on('notification') isn't fired
#### Platform and Version (eg. Android 5.0 or iOS 9.2.1)

Windows 8.1
#### (Android) What device vendor (e.g. Samsung, HTC, Sony...)

Microsoft Lumia 640
#### Cordova CLI version and cordova platform version

cordova --version      6.2.0
cordova platform version windows 4.3.2
#### Plugin version

1.7.0

It works totally fine when app is closed (on('notification') is fired with launchArgs), or when the app is open (on('notification') is fired with all data). But when the app is running but not on foreground and the user taps on a notification, on('notification') isn't fired at all.

Is this a bug, or expected behavior? Any way workaround available?

**Comment by rakatyal** [comment]
Yes this is expected behavior. In this scenario, the on-notification is fired when the notification is received and not on tap.

**Comment by mdbw3** [comment]
Thank you for your reply. The on-notification event isn't fired at all when the app is running in the background, not on tap, not on received. 

On both iOS and Android, the on receive event is fired when the user taps on a notification. It doesn't matter if the app is totally closed, or still running in the background. Why is this different for Windows? 

As a result you can't perform any action when a user taps on the notification, when the app is still running in the background. I would expect that the app would show up, and immediately after that the on receive event will be triggered. With the current behaviour, the app does show up, but nothing else happens. A user would expect that the action that belongs to the notification is fired after showing, right? That's how almost every app that uses push notifications work.

Please let me know if I'm missing something. Thank you!

**Comment by ajaysaini95700** [comment]
same problem in android. i want to open particular page whenever notification is clicked.

**Comment by code-constructor** [comment]
I had the same behaviour. The problem was the notification payload which i've sended to google. All informations should be in the "data" key, nothing in the "notification". Otherwise the OS preprocess the notification and not the plugin.

I hope this help!

**Comment by ajaysaini95700** [comment]
Thanks for reply

I am getting all information from "data" key. But I want to open specific
page whenever tapping on notification. So how to open specific page ??

**Comment by macdonst** [comment]
@ajaysaini95700 please open a different issue for this problem. It gets confusing when two different threads are in the same issue.

**Comment by mdbw3** [comment]
Thanks, @macdonst. For clarity I'll repost my question for @rakatyal 

> Thank you for your reply. The on-notification event isn't fired at all when the app is running in the background, not on tap, not on received.
> 
> On both iOS and Android, the on receive event is fired when the user taps on a notification. It doesn't matter if the app is totally closed, or still running in the background. Why is this different for Windows?
> 
> As a result you can't perform any action when a user taps on the notification, when the app is still running in the background. I would expect that the app would show up, and immediately after that the on receive event will be triggered. With the current behaviour, the app does show up, but nothing else happens. A user would expect that the action that belongs to the notification is fired after showing, right? That's how almost every app that uses push notifications work.
> 
> Please let me know if I'm missing something. Thank you!

**Comment by mdbw3** [comment]
After some debugging the on('notification') event is triggered even when app is running in background. There is one problem left: If you tap on a notification in notification centre while the app is open (in fore- or background), the app opens, but no event is fired. Steps to reproduce:

- Open app
- Hide app by pressing Windows button
- Send several notifications to device
- Open app (on('notification') events were fired while in background)
- Swipe form top of screen to bottom to open notification centre
- Tap on a notification
- -> Nothing happens, no event is fired

On both iOS and Android the on('notification') event is fired in this case.

Are you planning to fix this behavior?

**Comment by GitMarJansen** [comment]
I get the same result. On iOS, everything works fine, the on('notification') gets data when triggered from the notification area (dormant app). On Android the payload is not passed when the app was dormant. Looking at the GCM page (https://developers.google.com/cloud-messaging/concept-options) the notification should contain the visible part of the push notification. When passing everything in data, iOS does not work anymore. (so the solution above to put everything in data instead of notification is not valid)
I am using this plugin since february and am sure it worked correct at that moment.

**Comment by macdonst** [comment]
@GitMarJansen if you are using GCM on iOS you would send the following push payload:

```
{
    "notification": {
        "title": "AUX Scrum",
        "body": "Scrum: Daily touchbase @ 10am Please be on time so we can cover everything on the agenda."
    }
}
```

but on Android you would send: 

```
{
    "data": {
        "title": "AUX Scrum",
        "body": "Scrum: Daily touchbase @ 10am Please be on time so we can cover everything on the agenda."
    }
}
```

I am aware that the Android way doesn't follow the GCM docs exactly but it is they way it has to work for now. When you send a message using the 'notification' payload on Android the GCM listener never gets called. So that is why we need to used the 'data' payload to ensure the GCM listener gets called so we can fire the JS event.

**Comment by GitMarJansen** [comment]
Hmmm. The reason we use gcm is to get rid of platform specific implementations. Is this likely to be working in the near future?

**Comment by macdonst** [comment]
@GitMarJansen It won't change with a GCM based implementation. I haven't done a lot of investigation into FCM but I have hopes.

**Comment by JD76** [comment]
@mdbw3 -Hi, how did you find how to fire the event on('notification') when the app is in background? I tried in every way, but if the app is in background the event is never called. (I have the 1.6.3 version of the plugin because the 1.7.0 is giving me an error - issue #1036 )

**Comment by mdbw3** [comment]
I can see the on('notification') event in the Visual Studio Javascript Console when the app is in background.

@rakatyal could you please look into my question?

> There is one problem left: If you tap on a notification in notification centre while the app is running (in fore- or background), the app opens, but no event is fired. Steps to reproduce:
> 
> Open app
> Hide app by pressing Windows button
> Send several notifications to device
> Open app (on('notification') events were fired while in background)
> Swipe from top of screen to bottom to open notification centre
> Tap on a notification
> -> Nothing happens, no event is fired
> On both iOS and Android the on('notification') event is fired in this case.
> 
> Are you planning to fix this behavior?

**Comment by LightZam** [comment]
@macdonst 

check this

https://firebase.google.com/docs/cloud-messaging/concept-options#notifications_and_data_messages

the document explain the different between data and notification

**Comment by macdonst** [comment]
@LightZam thanks for the link. I do understand the differences between data and notification. What I need to test with FCM is if the app can get notified when notification type pushes arrive. Right now if you send a notification type push with GCM the GCM service we use is never called so we can't do all the fancy things like action buttons, background notifications that everyone wants.

**Comment by tyrtok** [comment]
@macdonst  @all 
what is the expected behaviour when app is running in foreground and a push notification arrive? 
a push notification msg displayed + onnotification event or only the onnotification event?

**Comment by mdbw3** [comment]
@rakatyal I'm still hoping for an answer on my question:

> There is one problem left: If you tap on a notification in notification centre while the app is running (in fore- or background), the app opens, but no event is fired. Steps to reproduce:
> 
> Open app
> Hide app by pressing Windows button
> Send several notifications to device
> Open app (on('notification') events were fired while in background)
> Swipe from top of screen to bottom to open notification centre
> Tap on a notification
> -> Nothing happens, no event is fired
> On both iOS and Android the on('notification') event is fired in this case.

Is this expected behavior? Due to this behavior, we can't use this plugin on any of our Windows projects, whereas the Android and iOS solution is behaving like I expect.


To all others: please open a new issue and do not use this issue for other problems or questions.

**Comment by daserge** [comment]
@mdbw3, I've tested this scenario and it works for me on Windows 8.1:
```javascript
    bindEvents: function() {
        document.addEventListener('deviceready', this.onDeviceReady, false);
        document.addEventListener('activated', this.onActivated, false);
    },
...
    onActivated: function (context) {
        var launchArgs = context ? context.args : null;
        if (launchArgs) {         //If present, app launched through push notification
            if (app.isReady) {
                // Background (suspended) mode case:
                app.registerForPush();
            } // else - coldStart case, will be handled in onDeviceReady
        }
    }
...
    onDeviceReady: function() {
        app.receivedEvent('deviceready');
        app.isReady = true;

        app.registerForPush();
    }
...
    registerForPush: function() {
        app.push = PushNotification.init({
            android: {
                senderID: "12345679"
            },
            browser: {
                pushServiceURL: 'http://push.api.phonegap.com/v1/push'
            },
            ios: {
                alert: "true",
                badge: "true",
                sound: "true"
            },
            windows: {}
        });

        app.push.on('registration', function (data) {
            // data.registrationId
            var msg = 'data.registrationId: ' + data.registrationId;
            navigator && navigator.notification && navigator.notification.alert(msg);
            console.log(msg);

            app.registrationId = data.registrationId;
        });

        app.push.on('notification', function (data) {
            // data.message,
            // data.title,
            // data.count,
            // data.sound,
            // data.image,
            // data.additionalData
            var msg = 'message: ' + data.message
                + '\ntitle: ' + data.title
                + '\ncount: ' + data.count
                + '\nsound: ' + data.sound
                + '\nimage: ' + data.image
                + '\nlaunchArgs: ' + data.launchArgs
                + '\nadditionalData: ' + JSON.stringify(data.additionalData);
            navigator && navigator.notification && navigator.notification.alert(msg);
            console.log(msg);
        });

        app.push.on('error', function (e) {
            // e.message
            var msg = 'error: ' + e.message;
            navigator && navigator.notification && navigator.notification.alert(msg);
            console.log(msg);
        });
    }
```

This way it works on tapping a notification when the app is suspended if it has `launch` attribute:
```xml
<?xml version="1.0" encoding="utf-8"?>
<toast launch="{&quot;myContext&quot;:&quot;12345&quot;}">
<visual><binding template="ToastText01">
<text id="1">Test message</text>
</binding>
</visual>
</toast>
```

I've faced some issues though:
1. An app does not wake up when a notification without `launch` attribute is tapped,
2. This does not work on Windows Phone 8.1 or Windows 10 (probably because of security/device registration server connection errors - investigating).

**Comment by daserge** [comment]
Windows 10 / Windows 10 Mobile works as well (note that https://github.com/apache/cordova-windows/pull/191 is required for Windows 10 Anniversary Update/RS1).
Windows Phone 8.1 also works after fixing Store association issues in package.phone.appxmanifest - VS did not update all the fields for some reason (fields `Identity Name`, `Publisher`).
Windows Phone 8.1 manifest should also be updated to set `Toast Capable = Yes` - I'm going to look why this did not update with `prepare` from `<preference name="WindowsToastCapable" value="true"/>`.

**Comment by daserge** [comment]
Regarding this:
> 1. An app does not wake up when a notification without launch attribute is tapped,

This issue occurs only for Windows 8.1 - we have white/frozen app window with VS debugger attached or just no `activated` or `on.notification` event otherwise.

Windows Phone 8.1 or Windows 10 does not have such issue.

**Comment by vikash28** [comment]
Yet after all this implement nothing over come it
???????

**Comment by mdbw3** [comment]
I suppose you're asking if there is a solution for the issues mentioned above? This is how I got it all working on Windows Phone:

- Listen to 'notification' event (of push notification plugin) to catch notifications when app is open
- Listen to 'activated' event to catch notifications which belong to tapped items of WP Notification Center

**Comment by tfrenken** [comment]
Hi all. I was wondering how @daserge code has to be changed in order to work with ionic 2. In general, the code works. The onActivated method is called, but the ionic2 context seems to be not available. References to "this" do not work. Any ideas on this?

**Comment by anjali14garg** [comment]
Thanks a lot @macdonst for your answer. It worked for me..

**Comment by mennocch** [comment]
If the user dismiss the notification instead click on it, the onNotification dont fire right? This is the expected behavior?
And if is the case, how to proced???
Tks in advance

**Comment by iranicus** [comment]
So I recently ran into this issue while the app is in the background and the strange thing is when I use Postman to send the notification to my device instead the logic works fine, the phone vibrates, sound plays, notification appears, and the on('notification') logic is executed when the user interacts with it. However I get no vibration or sound when another device sends the notification, with the on('notification') not executing after the user interacts with the notification and it opens the app up. Taking a look at the response I got when another device sent the notification:
{
   "sound":"default",
   "title":"testTitle",
   "message":"testMessage1",
   "additionalData":{
      "priority":"high",
      "sender":"143251",
      "coldstart":false,
      "foreground":true
   }
}
But when postman sends the notification (and I get the behavior we want) the notification details are:
{  
   "sound":"default",
   "title":"testTitle",
   "message":"testMessage1",
   "additionalData":{  
      "priority":"high",
      "sender":"143251",
      "dismissed":false,
      "coldstart":false,
      "foreground":false
   }
}
The difference I noticed being the "dismissed": false added which wasn't in the payload of the notification in Postman. Might be something.
